parquet-variant/src/main/java/org/apache/parquet/variant/VariantUtil.java

```diff
@@ -0,0 +1,843 @@
+      this.numElements = numElements;
+      this.idSize = idSize;
+      this.offsetSize = offsetSize;
+      this.idStart = idStart;
+      this.offsetStart = offsetStart;
+      this.dataStart = dataStart;
+    }
+  }
+
+  /**
+   * An interface for the Variant object handler.
+   * @param <T> The return type of the handler
+   */
+  public interface ObjectHandler<T> {
+    /**
+     * @param objectInfo The details of the Variant object
+     */
+    T apply(ObjectInfo objectInfo);
+  }
+
+  /**
+   * An interface for the Variant object handler.
+   * @param <T> The return type of the handler
+   */
+  public interface ObjectHandlerException<T> {
+    /**
+     * @param objectInfo The details of the Variant object
+     */
+    T apply(ObjectInfo objectInfo) throws IOException;
+  }
+
+  /**
+   * A helper function to access a Variant object, at `value[pos...]`.
+   * @param value The Variant value
+   * @param pos The starting index of the Variant value
+   * @param handler The handler to process the object
+   * @return The result of the handler
+   * @param <T> The return type of the handler
+   */
+  public static <T> T handleObject(byte[] value, int pos, ObjectHandler<T> handler) {
+    ObjectInfo info = parseObject(value, pos);
+    return handler.apply(info);
+  }
+
+  /**
+   * Same as `handleObject` but handler can throw IOException.
+   */
+  public static <T> T handleObjectException(byte[] value, int pos, ObjectHandlerException<T> handler)
+      throws IOException {
+    ObjectInfo info = parseObject(value, pos);
+    return handler.apply(info);
+  }
+
+  /**
+   * Parses the object at `value[pos...]`, and returns the object details.
+   */
+  private static ObjectInfo parseObject(byte[] value, int pos) {
+    checkIndex(pos, value.length);
+    int basicType = value[pos] & BASIC_TYPE_MASK;
+    int typeInfo = (value[pos] >> BASIC_TYPE_BITS) & PRIMITIVE_TYPE_MASK;
+    if (basicType != OBJECT) {
+      throw unexpectedType(Type.OBJECT);
+    }
+    // Refer to the comment of the `OBJECT` constant for the details of the object header encoding.
+    // Suppose `typeInfo` has a bit representation of 0_b4_b3b2_b1b0, the following line extracts
+    // b4 to determine whether the object uses a 1/4-byte size.
+    boolean largeSize = ((typeInfo >> 4) & 0x1) != 0;
+    int sizeBytes = (largeSize ? U32_SIZE : 1);
+    int numElements = readUnsigned(value, pos + 1, sizeBytes);
+    // Extracts b3b2 to determine the integer size of the field id list.
+    int idSize = ((typeInfo >> 2) & 0x3) + 1;
+    // Extracts b1b0 to determine the integer size of the offset list.
+    int offsetSize = (typeInfo & 0x3) + 1;
+    int idStart = pos + 1 + sizeBytes;
+    int offsetStart = idStart + numElements * idSize;
+    int dataStart = offsetStart + (numElements + 1) * offsetSize;
+    return new ObjectInfo(numElements, idSize, offsetSize, idStart, offsetStart, dataStart);
```

Comment:
I understand why you'd build `pos` into the offsets passed back, but returning the absolute position in the buffer means that the object info (or similarly, array info) is not applicable to values that are returned by `Variant#getValue` even though the bytes are the same. The same bytes at a different offset produce different `ObjectInfo` instances. This isn't a huge problem, but it seems like it could cause some bugs if callers decide to reuse any values they take from the object info.